OSDOCS-11461:support matrix module for  UDN (#81484)
openshift/openshift-docs: modules/nw-udn-support-matrix-primary-secondary.adoc

//module included in the following assembly:
//
// *networkking/multiple_networks/understanding-user-defined-networks.adoc

:_mod-docs-content-type: CONCEPT
[id="support-matrix-for-primary-secondary-networks_{context}"]
= Support matrix for primary and secondary networks

The following table outlines supported features of primary and secondary networks created by `UserDefinedNetwork` (UDN) and other secondary networks. For more information on, the limitations of user-defined networks see: xref:../networking/ . For more information on secondary networks that are created through network attachement defintions using multus, see xref:../networking/.

// column 4 and 8 will need put back in once supported
[cols="<.^2,^.^1,^.^1,^.^1,^.^1,^.^1,^.^1,^.^1,^.^1,^.^1,^.^1"]
|====
|Network type
|Definition
|Role
|east-west traffic
|north-south traffic
|Persistent IPs

> |`MultiNetworkPolicy`resource

Reviewer: ```suggestion
|`MultiNetworkPolicy` resource
```

This is not rendering correctly. I assume it's the missing space.

|Routes
|Services
|`NetworkPolicy` resource
|`EgressIP` resource

|User-defined network
|A network that is created by the user using the UDN custom resource definition (CRD) and declared as primary.
|Administrator and developer
|X `Layer2` and `Layer3` topology types
|X `Layer2` and `Layer3` topology types
|X `Layer2` topology type
|Not supported
|X `Layer2` and `Layer3` topology types
|X
|X
|X

|Secondary network
| An additional network presented to pods. Created using a Network Attachment Definition (NAD) custom resource.
|Administrator only

> |Not supported

Reviewer: East west traffic is fully supported for all topologies in secondary NAD.

//module included in the following assembly:
//
// *networkking/multiple_networks/understanding-user-defined-networks.adoc

:_mod-docs-content-type: CONCEPT
[id="support-matrix-for-primary-secondary-networks_{context}"]
= Support matrix for primary and secondary networks

> The following table outlines supported features of primary and secondary networks created by `UserDefinedNetwork` (UDN) and other secondary networks. For more information on, the limitations of user-defined networks see: xref:../networking/ . For more information on secondary networks that are created through network attachement defintions using multus, see xref:../networking/.

Reviewer 2: i'm guessing the xrefs are for the content in your related PRs?

[cols="<.^3,^.^1,^.^1,^.^1,^.^1,^.^1,^.^1,^.^1,^.^1,^.^1,^.^1"]
|====
|Network type |definition |role |services |east-west traffic |north-sourth traffic |Persistent IPs |`NetworkPolicy`| `MultiNetworkPolicy` |Ingress |Egress IP

> |UDN primary

Reviewer 3: This is for 4.17 right?  Then service and networkpolicy is not ready in 4.17.
Can you find some DEV to review firstly, thanks!

Author: Services and Netpol have been commented out.

> |A network that is created by the user using the UDN custom resource definition (CRD) and declared either primary or secondary.

Reviewer 2: based on this being called UDN primary i'd expect it to be declared as primary. is that right? if not, that's confusing! 😅 
```suggestion
|A network that is created by the user using the UDN custom resource definition (CRD) and declared as primary.
```

|Administrator and developer
|X [^1^] [^2^]
|X [^1^] [^2^]
|X [^1^] [^2^]
|X [^1^]
|X [^1^] [^2^]
|
|X [^1^] [^2^]
|

|UDN Secondary

> |A network that is created by the user using the UDN (CRD) and declared either primary or secondary.

Reviewer 2: see comment about UDN primary
```suggestion
|A network that is created by the user using the UDN custom resource definition (CRD) and declared as secondary.
```

|Administrator and developer
|
|X [^1^] [^2^]
|
|
|
|X
|
|

|Secondary network

> | An additional network presented to pods. Created using Network Attachment Definition (NAD) custom resource.

Reviewer 2: needs an article:
```suggestion
| An additional network presented to pods. Created using a Network Attachment Definition (NAD) custom resource.
```

//module included in the following assembly:
//
// *networkking/multiple_networks/understanding-user-defined-networks.adoc

:_mod-docs-content-type: CONCEPT
[id="support-matrix-for-primary-secondary-networks_{context}"]
= Support matrix for primary and secondary networks

The following table outlines supported features of primary and secondary networks created by `UserDefinedNetwork` (UDN) and other secondary networks. For more information on, the limitations of user-defined networks see: xref:../networking/ . For more information on secondary networks that are created through network attachement defintions using multus, see xref:../networking/.

// column 4 and 8 will need put back in once supported
[cols="<.^3,^.^1,^.^1,^.^1,^.^1,^.^1,^.^1,^.^1,^.^1"]
|====
|Network type
|definition
|role

> //| services

Author: QE above said `services` and `NetworkPolicy` is out for 4.17.0

> //| services

Reviewer 4: ack then why is `EgressIP` not commented out?

> //| services

Author: Because I wasn't told that was out and the planning document I shared still has it in.

> //| services

Reviewer 4: hmm pretty sure I informed you and steven: see https://redhat-internal.slack.com/archives/C077NGF5HE1/p1726126212570869?thread_ts=1725609712.540449&cid=C077NGF5HE1

> //| services

Author: I will comment it out.

|east-west traffic
|north-south traffic
|Persistent IPs

> //|`NetworkPolicy`

Reviewer 4: resolved by above comment

|`MultiNetworkPolicy`

> |Ingress

Reviewer: By ingress here, do we mean OpenShift [ingress](https://docs.openshift.com/container-platform/4.16/networking/configuring_ingress_cluster_traffic/configuring-ingress-cluster-traffic-ingress-controller.html#nw-using-ingress-and-routes_configuring-ingress-cluster-traffic-ingress-controller) ? If so, I have mislead you. We don't support that on secondary UDNs. 

I also don't understand the separation between a secondary UDN and a secondary network. These should be the same.

Author: @maiqueb , the matrix doesn't have ingress X for secondary networks created using UDN. The reason that we are separating out secondary UDN and secondary network is to demonstrate the difference between features for creating network segmentation using the UDN resource vs a NAD resource. 

These tables are tough to keep track of but I put a preview in the top comment: https://81484--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/multiple_networks/understanding-multiple-networks.html#support-matrix-for-primary-secondary-networks_understanding-multiple-networks

Reviewer: I get that; what I'm trying to say is I think the feature set on the secondary UDN is the same, independently of how it was created.

Author: copy, so I should add the same set of (x)s there too. Will do.

Author: @maiqueb updated. How's that look?

Reviewer: I think you're missing east / west traffic for secondary networks - I would define it as `X [1][3]` .

Also, `Ingress` (assuming its OpenShift Ingress / routes) is **not** supported for secondary networks. In *any* topology.

Finally, I find it confusing to mark multi-network policy as available on all topologies; are we indicating we now support L3 secondary networks ? IIUC, the documentation was added in 4.13 / 4.14 but never shown to the user. Hence, I would also define it as `X [1][3]` .

> |Ingress

Reviewer: Yeah, I have issues understanding this as well.

> |Ingress

Author: I'll change to routes.

|Egress IP

|UDN primary
|A network that is created by the user using the UDN custom resource definition (CRD) and declared as primary.
|Administrator and developer

> //|X [^1^] [^2^]

Reviewer: I actually assumed X to be supported on all topologies ...

> //|X [^1^] [^2^]

Author: x means supported

> //|X [^1^] [^2^]

Author: Added `Not supported` to clarify things.

|X [^1^] [^2^]
|X [^1^] [^2^]
|X [^1^]
//|X [^1^] [^2^]
|
|X [^1^] [^2^]
|

|UDN Secondary
|A network that is created by the user using the UDN (CRD) and declared as secondary.
|Administrator and developer
//|X

> |X [^1^] [^2^]

Reviewer 4: this is gonna get harder with the admin CRDs cause there we do support localnet or plan to then overlay will have support, would be good to note this down for 4.18 docs this needs to change

Author: The footnotes have `localnet` and I thought we could do CUDN as another row if necessary.

Author: Or perhaps we add CUDN as a footnote.

|X [^1^] [^2^]

> |X [^1^]

Reviewer 4: @maiqueb are persistent IPs not supported for L3?

Reviewer: No. Only L2 / localnet.

> //|

Reviewer 4: let's remove the commented out lines or move them to the end? I am having a really hard time reviewing this piece of change :( cause I am having to manually correlate the preview with the lines here..

Author: done

|
|X [^1^] [^2^]
|

> |Secondary network

Author: I don't see that in the draft document I shared on sept 5.

> |Secondary network

Reviewer 4: not sure who was taking notes :) ...

> |Secondary network

Author: This is still "secondary network" because https://github.com/openshift/openshift-docs/pull/81740 didn't merge. In 81740 we explicitly call out NADs secondary. However, we don't in our current docs. To achieve what we were trying to do I wrote in the explanations.
